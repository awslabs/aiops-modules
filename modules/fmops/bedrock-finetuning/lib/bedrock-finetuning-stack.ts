import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as iam from "aws-cdk-lib/aws-iam";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as tasks from "aws-cdk-lib/aws-stepfunctions-tasks";
import * as sfn from "aws-cdk-lib/aws-stepfunctions";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as events from "aws-cdk-lib/aws-events";
import * as targets from "aws-cdk-lib/aws-events-targets";
import * as kms from "aws-cdk-lib/aws-kms";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as logs from "aws-cdk-lib/aws-logs";
import { NagSuppressions } from "cdk-nag";

interface AmazonBedrockFinetuningStackProps extends cdk.StackProps {
  projectName?: string;
  deploymentName?: string;
  moduleName?: string;
  bucketName?: string;
  removalPolicy: "DESTROY" | "RETAIN";
  bedrockBaseModelID: string;
  vpcId?: string;
  subnetIds: string[];
}

export class AmazonBedrockFinetuningStack extends cdk.Stack {
  bucketName: string;

  constructor(
    scope: Construct,
    id: string,
    props: AmazonBedrockFinetuningStackProps,
  ) {
    super(scope, id, props);

    // create S3 bucket
    const inputBucket = this.getBucket(props);
    this.bucketName = inputBucket.bucketName;

    // Create a KMS Key
    const key = new kms.Key(this, "MyKey", {
      enableKeyRotation: true,
      description: "Key for encrypting Bedrock fine-tuning jobs",
    });

    // creating IAM role to pass to Amazon Bedrock
    const bedrockS3PolicyStatement = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ["s3:GetObject", "s3:PutObject", "s3:ListBucket"],
      resources: [inputBucket.bucketArn, `${inputBucket.bucketArn}/*`],
      conditions: {
        StringEquals: {
          "aws:PrincipalAccount": this.account,
        },
      },
    });
    const bedrockPassRolePolicyStatement = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ["iam:PassRole"],
      resources: ["*"],
      conditions: {
        StringEquals: {
          "aws:PrincipalAccount": this.account,
        },
      },
    });
    const KMSKeyPolicyStatement = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: [
        "kms:Encrypt",
        "kms:ReEncrypt*",
        "kms:GenerateDataKey*",
        "kms:Decrypt",
        "kms:DescribeKey",
      ],
      resources: [key.keyArn],
      conditions: {
        StringEquals: {
          "aws:PrincipalAccount": this.account,
        },
      },
    });

    const bedrockPolicy = new iam.Policy(this, "CustomPolicy", {
      statements: [bedrockS3PolicyStatement, bedrockPassRolePolicyStatement],
    });
    const bedrockRole = new iam.Role(this, "MyRole", {
      assumedBy: new iam.CompositePrincipal(
        new iam.ServicePrincipal("bedrock.amazonaws.com")
          .withConditions({
            StringEquals: {
              "aws:SourceAccount": this.account,
            },
          })
          .withConditions({
            ArnEquals: {
              "aws:SourceArn": `arn:${this.partition}:bedrock:${this.region}:${this.account}:model-customization-job/*`,
            },
          }),
      ),
    });
    bedrockRole.attachInlinePolicy(bedrockPolicy);

    const vpc = ec2.Vpc.fromLookup(this, "MyVpc", {
      vpcId: props.vpcId, // Replace with your VPC ID
    });

    //create security group, add configuration depending on your use case and setup
    const securityGroup = new ec2.SecurityGroup(this, "SecurityGroup", {
      vpc: vpc,
      description: "Allow inbound traffic",
    });

    // creating lambda function for running the model finetuning job
    const modelFinetuningLambda = new lambda.Function(
      this,
      "ObjectDetectionLambda",
      {
        runtime: lambda.Runtime.PYTHON_3_12,
        code: lambda.Code.fromAsset("src/lambda-functions"),
        timeout: cdk.Duration.seconds(60),
        memorySize: 512,
        handler: "modelFinetuningLambdaCode.lambda_handler",
        functionName: "modelFinetuningLambdaFcn",
        environment: {
          role_arn: bedrockRole.roleArn,
          kms_key_id: key.keyId,
          base_model_id: props.bedrockBaseModelID,
          vpc_subnets: JSON.stringify(props.subnetIds),
          vpc_sec_group: securityGroup.uniqueId,
        },
        vpc: vpc,
      },
    );

    modelFinetuningLambda.addToRolePolicy(
      new iam.PolicyStatement({
        actions: ["bedrock:CreateModelCustomizationJob"],
        resources: ["*"],
        conditions: {
          StringEquals: {
            "aws:PrincipalAccount": this.account,
          },
        },
      }),
    );
    modelFinetuningLambda.addToRolePolicy(
      new iam.PolicyStatement({
        actions: ["iam:PassRole"],
        resources: [`${bedrockRole.roleArn}`],
        conditions: {
          StringEquals: {
            "aws:PrincipalAccount": this.account,
          },
        },
      }),
    );
    modelFinetuningLambda.addToRolePolicy(KMSKeyPolicyStatement);
    // creating step function that will trigger the lambda function
    const modelFinetuningLambdaTask = new tasks.LambdaInvoke(
      this,
      "ModelFinetuningLambdaTask",
      {
        lambdaFunction: modelFinetuningLambda,
        outputPath: "$.Payload",
      },
    );
    const definition = modelFinetuningLambdaTask.next(
      new sfn.Succeed(this, "Succeed"),
    );
    const stateMachine = new sfn.StateMachine(this, "MyStateMachine", {
      definitionBody: sfn.DefinitionBody.fromChainable(definition),
      timeout: cdk.Duration.minutes(5),
      stateMachineName: "BedrockFinetuning",
      tracingEnabled: true,
      logs: {
        destination: new logs.LogGroup(this, "MyLogGroup", {
          logGroupName: "/aws/vendedlogs/states/BedrockFinetuning",
          removalPolicy: cdk.RemovalPolicy.DESTROY,
          retention: logs.RetentionDays.ONE_MONTH,
        }),
        level: sfn.LogLevel.ALL,
      },
    });

    // Grant the state machine permission to invoke the Lambda function
    modelFinetuningLambda.grantInvoke(stateMachine.role);

    const rule = new events.Rule(this, "Rule", {
      eventPattern: {
        source: ["aws.s3"],
        detailType: ["Object Created"],
        detail: {
          bucket: {
            name: [this.bucketName],
          },
        },
      },
    });

    rule.addTarget(new targets.SfnStateMachine(stateMachine));

    NagSuppressions.addResourceSuppressions(
      [inputBucket],
      [
        {
          id: "AwsSolutions-S1",
          reason: "Bucket logging is not required",
        },
      ],
      true,
    );
    NagSuppressions.addStackSuppressions(this, [
      {
        id: "AwsSolutions-IAM4",
        reason: "This is a necessary managed policy for Lambda execution",
      },
    ]);
    NagSuppressions.addStackSuppressions(
      this,
      [
        {
          id: "AwsSolutions-IAM5",
          reason: "star required as default actions are not working",
        },
      ],
      true,
    );
  }

  private getBucket(props: AmazonBedrockFinetuningStackProps): s3.IBucket {
    if (props.bucketName) {
      return s3.Bucket.fromBucketName(this, "BedrockBucket", props.bucketName);
    } else {
      return new s3.Bucket(this, "BedrockBucket", {
        bucketName: `${props.deploymentName}-${props.moduleName}-${this.account}`,
        removalPolicy:
          props.removalPolicy == "DESTROY"
            ? cdk.RemovalPolicy.DESTROY
            : cdk.RemovalPolicy.RETAIN,
        autoDeleteObjects: props.removalPolicy == "DESTROY",
        eventBridgeEnabled: true,
        enforceSSL: true,
        encryption: s3.BucketEncryption.KMS_MANAGED,
      });
    }
  }
}
