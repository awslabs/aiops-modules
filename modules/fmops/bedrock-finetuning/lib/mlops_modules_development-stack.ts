import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as iam from "aws-cdk-lib/aws-iam";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as tasks from "aws-cdk-lib/aws-stepfunctions-tasks";
import * as sfn from "aws-cdk-lib/aws-stepfunctions";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as events from "aws-cdk-lib/aws-events";
import * as targets from "aws-cdk-lib/aws-events-targets";
import * as kms from "aws-cdk-lib/aws-kms";

export class MlopsModulesDevelopmentStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // create S3 bucket
    const inputBucket = new s3.Bucket(this, "BedrockBucket", {
      bucketName: "bedrock-input-data-lj",
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
      eventBridgeEnabled: true,
    });

    // Create a KMS Key
    const key = new kms.Key(this, "MyKey", {
      enableKeyRotation: true,
      description: "Key for encrypting Bedrock fine-tuning jobs",
    });

    // creating IAM role to pass to Amazon Bedrock
    const bedrockS3PolicyStatement = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ["s3:GetObject", "s3:PutObject", "s3:ListBucket"],
      resources: [inputBucket.bucketArn, `${inputBucket.bucketArn}/*`],
      conditions: {
        StringEquals: {
          "aws:PrincipalAccount": this.account,
        },
      },
    });
    const bedrockPassRolePolicyStatement = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ["iam:PassRole"],
      resources: ["*"],
      conditions: {
        StringEquals: {
          "aws:PrincipalAccount": this.account,
        },
      },
    });
    const KMSKeyPolicyStatement = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ["kms:*"],
      resources: [key.keyArn],
      conditions: {
        StringEquals: {
          "aws:PrincipalAccount": this.account,
        },
      },
    });

    const bedrockPolicy = new iam.Policy(this, "CustomPolicy", {
      statements: [bedrockS3PolicyStatement, bedrockPassRolePolicyStatement],
    });
    const bedrockRole = new iam.Role(this, "MyRole", {
      assumedBy: new iam.CompositePrincipal(
        new iam.ServicePrincipal("bedrock.amazonaws.com")
          .withConditions({
            StringEquals: {
              "aws:SourceAccount": this.account,
            },
          })
          .withConditions({
            ArnEquals: {
              "aws:SourceArn": `arn:aws:bedrock:${this.region}:${this.account}:model-customization-job/*`,
            },
          }),
      ),
    });
    bedrockRole.attachInlinePolicy(bedrockPolicy);

    // creating lambda function for running the model finetuning job
    const modelFinetuningLambda = new lambda.Function(
      this,
      "ObjectDetectionLambda",
      {
        runtime: lambda.Runtime.PYTHON_3_11,
        code: lambda.Code.fromAsset("src/lambda-functions"),
        timeout: cdk.Duration.seconds(60),
        memorySize: 512,
        handler: "modelFinetuningLambdaCode.lambda_handler",
        functionName: "modelFinetuningLambdaFcn",
        environment: {
          role_arn: bedrockRole.roleArn,
          kms_key_id: key.keyId,
          base_model_id: "amazon.titan-text-express-v1:0:8k",
        },
      },
    );

    modelFinetuningLambda.addToRolePolicy(
      new iam.PolicyStatement({
        actions: ["bedrock:CreateModelCustomizationJob"],
        resources: ["*"],
        conditions: {
          StringEquals: {
            "aws:PrincipalAccount": this.account,
          },
        },
      }),
    );
    modelFinetuningLambda.addToRolePolicy(
      new iam.PolicyStatement({
        actions: ["iam:PassRole"],
        resources: [`${bedrockRole.roleArn}`],
        conditions: {
          StringEquals: {
            "aws:PrincipalAccount": this.account,
          },
        },
      }),
    );
    modelFinetuningLambda.addToRolePolicy(KMSKeyPolicyStatement);
    // creating step function that will trigger the lambda function
    const modelFinetuningLambdaTask = new tasks.LambdaInvoke(
      this,
      "ModelFinetuningLambdaTask",
      {
        lambdaFunction: modelFinetuningLambda,
        outputPath: "$.Payload",
      },
    );
    const definition = modelFinetuningLambdaTask.next(
      new sfn.Succeed(this, "Succeed"),
    );
    const stateMachine = new sfn.StateMachine(this, "MyStateMachine", {
      definition: definition,
      timeout: cdk.Duration.minutes(5),
      stateMachineName: "BedrockFinetuning",
    });

    // Grant the state machine permission to invoke the Lambda function
    modelFinetuningLambda.grantInvoke(stateMachine.role);

    const rule = new events.Rule(this, "Rule", {
      eventPattern: {
        source: ["aws.s3"],
        detailType: ["Object Created"],
        detail: {
          bucket: {
            name: ["bedrock-input-data-lj"],
          },
        },
      },
    });

    rule.addTarget(new targets.SfnStateMachine(stateMachine));
  }
}
