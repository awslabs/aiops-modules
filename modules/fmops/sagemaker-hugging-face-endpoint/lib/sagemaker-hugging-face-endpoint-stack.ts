import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as iam from "aws-cdk-lib/aws-iam";
import * as secrets from "aws-cdk-lib/aws-secretsmanager";
import * as sagemaker from "aws-cdk-lib/aws-sagemaker";
import {
  DeepLearningContainerImage,
  HuggingFaceSageMakerEndpoint,
  SageMakerInstanceType,
} from "@cdklabs/generative-ai-cdk-constructs";
import * as cdk_nag from "cdk-nag";

interface SagemakerHuggingFaceEndpointStackProps extends cdk.StackProps {
  projectName?: string;
  deploymentName?: string;
  moduleName?: string;
  huggingFaceModelID: string;
  instanceType: string;
  deepLearningContainerImage: string;
  vpcId?: string | undefined;
  subnetIds: string[];
  hfTokenSecretName?: string;
}

export class SagemakerHuggingFaceEndpointStack extends cdk.Stack {
  huggingFaceEndpoint: HuggingFaceSageMakerEndpoint;
  role: iam.Role;

  constructor(scope: Construct, id: string, props: SagemakerHuggingFaceEndpointStackProps) {
    super(scope, id, props);

    this.role = new iam.Role(this, "Role", {
      assumedBy: new iam.ServicePrincipal("sagemaker.amazonaws.com"),
      inlinePolicies: {
        inline: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                "ecr:BatchCheckLayerAvailability",
                "ecr:BatchGetImage",
                "ecr:Describe*",
                "ecr:GetAuthorizationToken",
                "ecr:GetDownloadUrlForLayer",
                "ec2:CreateNetworkInterface",
                "ec2:CreateNetworkInterfacePermission",
                "ec2:CreateVpcEndpoint",
                "ec2:DeleteNetworkInterface",
                "ec2:DeleteNetworkInterfacePermission",
                "ec2:Describe*",
                "logs:CreateLogDelivery",
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:DeleteLogDelivery",
                "logs:Describe*",
                "logs:GetLogDelivery",
                "logs:GetLogEvents",
                "logs:ListLogDeliveries",
                "logs:PutLogEvents",
                "logs:PutResourcePolicy",
                "logs:UpdateLogDelivery",
              ],
              resources: ["*"],
            }),
          ],
        }),
      },
    });

    let vpcConfig: sagemaker.CfnModel.VpcConfigProperty | undefined = undefined;

    if (props.vpcId) {
      const vpc = ec2.Vpc.fromLookup(this, "Vpc", { vpcId: props.vpcId });
      const securityGroup = new ec2.SecurityGroup(this, "SecurityGroup", {
        vpc: vpc,
      });
      vpcConfig = {
        securityGroupIds: [securityGroup.securityGroupId],
        subnets: props.subnetIds,
      };
    }

    const [containerImageRepoName, containerImageTag] = SagemakerHuggingFaceEndpointStack.parseContainerImage(
      props.deepLearningContainerImage,
    );

    const hfTokenEnvironmentVars = this.getSecretTokenVars(props.hfTokenSecretName);

    this.huggingFaceEndpoint = new HuggingFaceSageMakerEndpoint(this, "HuggingFace Endpoint", {
      modelId: props.huggingFaceModelID,
      instanceType: SageMakerInstanceType.of(props.instanceType),
      container: DeepLearningContainerImage.fromDeepLearningContainerImage(containerImageRepoName, containerImageTag),
      role: this.role,
      vpcConfig: vpcConfig,
      environment: {
        ...hfTokenEnvironmentVars,
      },
    });

    cdk_nag.NagSuppressions.addResourceSuppressions(
      this.role,
      [
        {
          id: "AwsSolutions-IAM5",
          reason: "Resource access restricted to S3 buckets (with a prefix) and ECR images",
        },
      ],
      true,
    );
  }

  static parseContainerImage(deepLearningContainerImage: string): [string, string] {
    const parts = deepLearningContainerImage.split(":");
    if (parts.length !== 2) {
      throw new Error(`Invalid container image: ${deepLearningContainerImage}`);
    }

    return [parts[0], parts[1]];
  }

  getSecretTokenVars(tokenSecretName: string | undefined): { [key: string]: string } {
    if (!tokenSecretName) {
      return {};
    }

    const secret = secrets.Secret.fromSecretNameV2(this, "HFTokenSecret", tokenSecretName);
    const token = secret.secretValueFromJson("TOKEN").toString();
    return {
      HUGGING_FACE_HUB_TOKEN: token,
    };
  }
}
