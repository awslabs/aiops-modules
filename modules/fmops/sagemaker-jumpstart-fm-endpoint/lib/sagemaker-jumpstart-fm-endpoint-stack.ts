import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import { Vpc, SecurityGroup } from "aws-cdk-lib/aws-ec2";
import { Effect, PolicyDocument, PolicyStatement, Role, ServicePrincipal } from "aws-cdk-lib/aws-iam";
import { CfnModel } from "aws-cdk-lib/aws-sagemaker";
import {
  JumpStartModel,
  SageMakerInstanceType,
  JumpStartSageMakerEndpoint,
} from "@cdklabs/generative-ai-cdk-constructs";
import * as cdk_nag from "cdk-nag";

interface SagemakerJumpStartFmEndpointStackProps extends cdk.StackProps {
  projectName?: string;
  deploymentName?: string;
  moduleName?: string;
  jumpStartModelName: string;
  instanceType: string;
  vpcId?: string | undefined;
  subnetIds: string[];
}

export class SagemakerJumpStartFmEndpointStack extends cdk.Stack {
  jumpStartEndpoint: JumpStartSageMakerEndpoint;
  role: Role;

  constructor(scope: Construct, id: string, props: SagemakerJumpStartFmEndpointStackProps) {
    super(scope, id, props);

    this.role = new Role(this, "Role", {
      assumedBy: new ServicePrincipal("sagemaker.amazonaws.com"),
      // managedPolicies: [
      //   ManagedPolicy.fromAwsManagedPolicyName("AmazonSageMakerFullAccess"),
      // ],
      inlinePolicies: {
        inline: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["s3:GetObject", "s3:ListBucket"],
              resources: [`arn:${this.partition}:s3:::jumpstart-*`],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                "ecr:BatchCheckLayerAvailability",
                "ecr:BatchGetImage",
                "ecr:Describe*",
                "ecr:GetAuthorizationToken",
                "ecr:GetDownloadUrlForLayer",
                "ec2:CreateNetworkInterface",
                "ec2:CreateNetworkInterfacePermission",
                "ec2:CreateVpcEndpoint",
                "ec2:DeleteNetworkInterface",
                "ec2:DeleteNetworkInterfacePermission",
                "ec2:Describe*",
                "logs:CreateLogDelivery",
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:DeleteLogDelivery",
                "logs:Describe*",
                "logs:GetLogDelivery",
                "logs:GetLogEvents",
                "logs:ListLogDeliveries",
                "logs:PutLogEvents",
                "logs:PutResourcePolicy",
                "logs:UpdateLogDelivery",
              ],
              resources: ["*"],
            }),
          ],
        }),
      },
    });

    let vpcConfig: CfnModel.VpcConfigProperty | undefined = undefined;

    if (props.vpcId) {
      const vpc = Vpc.fromLookup(this, "Vpc", { vpcId: props.vpcId });
      const securityGroup = new SecurityGroup(this, "SecurityGroup", {
        vpc: vpc,
      });
      vpcConfig = {
        securityGroupIds: [securityGroup.securityGroupId],
        subnets: props.subnetIds,
      };
    }

    this.jumpStartEndpoint = new JumpStartSageMakerEndpoint(this, "JumpStartEndpoint", {
      model: JumpStartModel.of(props.jumpStartModelName),
      instanceType: SageMakerInstanceType.of(props.instanceType),
      role: this.role,
      vpcConfig: vpcConfig,
    });

    cdk_nag.NagSuppressions.addResourceSuppressions(this.role, [
      {
        id: "AwsSolutions-IAM5",
        reason: "Resource access restricted to S3 buckets (with a prefix) and ECR images",
      },
    ]);
  }
}
