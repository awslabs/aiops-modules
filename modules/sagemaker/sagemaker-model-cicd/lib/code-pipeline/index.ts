export * from './mlops-code-pipeline-stack';
export * from './mlops-code-pipeline-stack-props';
export * from './model-build-support/model-build-support-stack';
export * from './model-build-support/model-build-support-stage';
export * from './model-build/model-build-code-pipeline-stack';
export * from './model-build/model-build-code-pipeline-stage';
export * from './model-deploy/model-deploy-code-pipeline-stack';
export * from './model-deploy/model-deploy-code-pipeline-stage';
export * from './model-endpoint/model-endpoint-stack';
export * from './model-endpoint/model-endpoint-stage';
