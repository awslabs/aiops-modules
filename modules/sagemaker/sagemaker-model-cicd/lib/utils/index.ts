export * from './create-resource-name';
export * from './get-seedfarmer-env-vars';
export * from './pipeline-artifacts-bucket';
