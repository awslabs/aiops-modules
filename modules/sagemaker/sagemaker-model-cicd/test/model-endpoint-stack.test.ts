import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { ModelEndpointStack } from '../lib';

test('ModelEndpointStack to match snapshot', () => {
  // timestamp is used to generate unique names for resources in this stack
  jest.spyOn(Date, 'now').mockReturnValue(123456789);
  const app = new cdk.App();
  const stack = new ModelEndpointStack(app, 'MyTestStack', {
    projectName: 'test-project',
    modelRegistryEnv: {
      name: 'model-registry',
      account: '123456789012',
      region: 'us-east-1',
      type: 'dev',
    },
    sagemakerArtifactsBucketName: 'sagemaker-artifacts-bucket',
    deployEnv: {
      name: 'deploy-env',
      account: '123456789013',
      type: 'dev',
    },
    modelPackageGroupName: 'model-package-group',
    deployModelInfo: {
      modelPackageArn: 'model-package-arn',
      modelDataUrl: 'model-data-url',
      imageUri: 'image-uri',
    },
  });
  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
  jest.restoreAllMocks();
});
