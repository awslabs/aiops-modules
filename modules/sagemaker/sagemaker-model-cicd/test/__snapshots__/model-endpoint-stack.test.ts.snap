// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelEndpointStack to match snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "SsmParameterValuetestprojectvpcsgidC96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Default": "/test-project/vpc/sg/id",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::SecurityGroup::Id>",
    },
    "SsmParameterValuetestprojectvpcsubnetsprivateidsC96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Default": "/test-project/vpc/subnets/private/ids",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
  },
  "Resources": {
    "Endpoint": {
      "DependsOn": [
        "EndpointConfig",
      ],
      "Properties": {
        "EndpointConfigName": "test-project-ec-123456789",
        "EndpointName": "test-project-e",
      },
      "Type": "AWS::SageMaker::Endpoint",
    },
    "EndpointConfig": {
      "DependsOn": [
        "Model",
      ],
      "Properties": {
        "EndpointConfigName": "test-project-ec-123456789",
        "KmsKeyId": {
          "Ref": "EndpointKmsKey6747A8D9",
        },
        "ProductionVariants": [
          {
            "InitialInstanceCount": 1,
            "InitialVariantWeight": 1,
            "InstanceType": "ml.m5.xlarge",
            "VariantName": "AllTraffic",
          },
        ],
      },
      "Type": "AWS::SageMaker::EndpointConfig",
    },
    "EndpointKmsKey6747A8D9": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Description": "Key used for encryption of data in Amazon SageMaker Endpoint",
        "EnableKeyRotation": true,
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Delete",
    },
    "Model": {
      "Properties": {
        "Containers": [
          {
            "ModelPackageName": "model-package-arn",
          },
        ],
        "ExecutionRoleArn": {
          "Fn::GetAtt": [
            "ModelExecutionRole3247A550",
            "Arn",
          ],
        },
        "ModelName": "model-package-group-123456789",
        "VpcConfig": {
          "SecurityGroupIds": [
            {
              "Ref": "SsmParameterValuetestprojectvpcsgidC96584B6F00A464EAD1953AFF4B05118Parameter",
            },
          ],
          "Subnets": {
            "Ref": "SsmParameterValuetestprojectvpcsubnetsprivateidsC96584B6F00A464EAD1953AFF4B05118Parameter",
          },
        },
      },
      "Type": "AWS::SageMaker::Model",
    },
    "ModelExecutionPolicy67130DDE": {
      "Properties": {
        "Description": "",
        "Path": "/",
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "kms:Encrypt",
                "kms:Decrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
                "kms:CreateGrant",
                "kms:ListGrants",
                "kms:DescribeKey",
              ],
              "Condition": {
                "ForAnyValue:StringEquals": {
                  "kms:ResourceAliases": "alias/sagemaker-artifacts-bucket",
                },
                "StringEquals": {
                  "kms:ViaService": "s3.us-east-1.amazonaws.com",
                },
              },
              "Effect": "Allow",
              "Resource": "arn:aws:kms:us-east-1:123456789012:key/*",
              "Sid": "ModelArtifactsBucketKeyAccess",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "ModelExecutionRole3247A550": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "sagemaker.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Ref": "ModelExecutionPolicy67130DDE",
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonSageMakerFullAccess",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ModelExecutionRoleDefaultPolicyFA1BB405": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::sagemaker-artifacts-bucket",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::sagemaker-artifacts-bucket/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ModelExecutionRoleDefaultPolicyFA1BB405",
        "Roles": [
          {
            "Ref": "ModelExecutionRole3247A550",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
